### PR TITLE
{lang}[foss/2022a] numba v0.56.4 w/ Python 3.10.4

### PR DESCRIPTION
**Pull request by Micket**
(created using `eb --new-pr`)

### CODE REVIEW

File: easybuild/easyconfigs/n/numba/numba-0.56.4-foss-2022a-CUDA-11.7.0.eb [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+use_pip = True
+sanity_pip_check = True
+
+local_llvmlite_preinstallopts = "export LLVM_CONFIG=${EBROOTLLVM}/bin/llvm-config && "
+local_llvmlite_preinstallopts += "export LLVMLITE_SKIP_LLVM_VERSION_CHECK=1 && "
+
+exts_list = [
+    ('llvmlite', '0.39.1', {
+        'patches': ['llvmlite-0.37.0_fix-LLVM-12.patch'],
+        'preinstallopts': "export LLVM_CONFIG=${EBROOTLLVM}/bin/llvm-config && export LLVMLITE_SKIP_LLVM_VERSION_CHECK=1 && ",
```
Comment (branfosj) [inline easybuild/easyconfigs/n/numba/numba-0.56.4-foss-2022a-CUDA-11.7.0.eb:31]:
```suggestion
        'preinstallopts': local_llvmlite_preinstallopts,
```

**Comment by Micket** [comment]
Test report by @Micket
**FAILED**
Build succeeded for 0 out of 1 (1 easyconfigs in total)
alvis-c1 - Linux Rocky Linux 8.5, x86_64, Intel Xeon Processor (Skylake), Python 3.6.8
See https://gist.github.com/4eb54075f9ff037c55c5ffb5980550a6 for a full test report.

**Comment by surak** [comment]
I made it for llvm12, with the following patches: 

```patch
--- ffi/passmanagers.cpp.orig	2022-12-06 15:41:01.230034402 +0000
+++ ffi/passmanagers.cpp	2022-12-06 15:41:19.678191543 +0000
@@ -17,9 +17,6 @@
 #include "llvm-c/Transforms/IPO.h"
 #include "llvm-c/Transforms/Scalar.h"
 #include "llvm/IR/LegacyPassManager.h"
-#if LLVM_VERSION_MAJOR > 11
-#include "llvm/IR/RemarkStreamer.h"
-#endif
 #include "llvm/IR/LLVMRemarkStreamer.h"
 #include "llvm/Remarks/RemarkStreamer.h"
 #include "llvm/Transforms/IPO.h"
 ```
 
 ```patch
--- ffi/targets.cpp.orig	2022-12-06 15:44:43.191925420 +0000
+++ ffi/targets.cpp	2022-12-06 15:44:53.916016793 +0000
@@ -204,7 +204,6 @@
         rm = Reloc::DynamicNoPIC;
 
     TargetOptions opt;
-    opt.PrintMachineCode = PrintMC;
     opt.MCOptions.ABIName = ABIName;
 
     bool jit = JIT;
```

```patch
--- llvmlite/tests/test_binding.py.orig	2022-12-06 18:00:21.729557505 +0000
+++ llvmlite/tests/test_binding.py	2022-12-06 18:00:49.641795081 +0000
@@ -18,6 +18,16 @@
 from llvmlite.tests import TestCase
 
 
+def clean_string_whitespace(x: str) -> str:
+    # Remove trailing whitespace from the end of each line
+    x = re.sub(r"\s+$", "", x, flags=re.MULTILINE)
+    # Remove intermediate blank lines
+    x = re.sub(r"\n\s*\n", r"\n", x, flags=re.MULTILINE)
+    # Remove extraneous whitespace from the beginning and end of the string
+    x = x.strip()
+    return x
+
+
 # arvm7l needs extra ABI symbols to link successfully
 if platform.machine() == 'armv7l':
     llvm.load_library_permanently('libgcc_s.so.1')
@@ -555,7 +565,10 @@
         bd = ir.IRBuilder(fn.append_basic_block(name="<>!*''#"))
         bd.ret(ir.Constant(ir.IntType(32), 12345))
         asm = str(mod)
-        self.assertEqual(asm, asm_nonalphanum_blocklabel)
+        self.assertEqual(
+            clean_string_whitespace(asm),
+            clean_string_whitespace(asm_nonalphanum_blocklabel)
+        )
 
     def test_global_context(self):
         gcontext1 = llvm.context.get_global_context()
@@ -640,7 +653,7 @@
     def test_version(self):
         major, minor, patch = llvm.llvm_version_info
         # one of these can be valid
-        valid = [(11,)]
+        valid = [(11,), (12,), (13,)]
         self.assertIn((major,), valid)
         self.assertIn(patch, range(10))
```


And this is the easyconfig. I don't know what's the official position on LLVM, but I need this fast, so I will use LLVM12: 


```python
easyblock = 'PythonBundle'

name = 'numba'
version = '0.56.4'
versionsuffix = '-CUDA-%(cudaver)s'

homepage = 'https://numba.pydata.org/'
description = """Numba is an Open Source NumPy-aware optimizing compiler for
Python sponsored by Continuum Analytics, Inc. It uses the remarkable LLVM
compiler infrastructure to compile Python syntax to machine code."""

toolchain = {'name': 'foss', 'version': '2022a'}
toolchainopts = {'pic': True}

dependencies = [
    ('Python', '3.10.4'),
    ('SciPy-bundle', '2022.05'),
    ('LLVM', '12.0.1'),
    ('CUDA', '11.7.0', '', SYSTEM),
]

use_pip = True
sanity_pip_check = True

local_llvmlite_preinstallopts = "export LLVM_CONFIG=${EBROOTLLVM}/bin/llvm-config && "
local_llvmlite_preinstallopts += "export LLVMLITE_SKIP_LLVM_VERSION_CHECK=1 && "


exts_list = [
    ('llvmlite', '0.39.1', {
        'preinstallopts': local_llvmlite_preinstallopts,
        'patches': ['llvmlite-0.39_passmanager.patch',
                    'llvmlite-0.39_target.patch',
                    'llvmlite-0.39_test-binding.patch',
        ],
    }),
    (name, version, {
        # 'patches': ['numba-0.54.1_fix-numpy-1.21.patch'],
        'checksums': [
            {'numba-0.56.4.tar.gz': '32d9fef412c81483d7efe0ceb6cf4d3310fde8b624a9cecca00f790573ac96ee'},
            # {'numba-0.54.1_fix-numpy-1.21.patch': 'f038e7adf0cd60f6bdde219b326aad4224fef46515799578ebbbdffa73224199'},
        ],
    }),
]

fix_python_shebang_for = ['bin/*']

sanity_check_paths = {
    'files': ['bin/numba', 'bin/pycc'],
    'dirs': ['lib/python%(pyshortver)s/site-packages'],
}

sanity_check_commands = [
    "python -m llvmlite.tests",
    "numba --help",
]

moduleclass = 'lang'

```

**Comment by Micket** [comment]
I don't think there is a reason to split up the existing patch into 3. 

I think i'd rather just wait for https://github.com/numba/llvmlite/pull/830

**Comment by surak** [comment]
> I don't think there is a reason to split up the existing patch into 3.
> 
> I think i'd rather just wait for [numba/llvmlite#830](https://github.com/numba/llvmlite/pull/830)

The patch was not applying. I generated them again - it was not just a split of the file.

**Comment by surak** [comment]
> I don't think there is a reason to split up the existing patch into 3.
> 
> I think i'd rather just wait for [numba/llvmlite#830](https://github.com/numba/llvmlite/pull/830)

I patched it for LLVM 14 and I have it running here. I can do a new MR, if you want

**Comment by boegel** [comment]
closing this since it's superseded by #16814

**Pull request closed without merging**